railties/lib/rails/generators/rails/app/templates/config/puma.rb.tt

```diff
@@ -2,38 +2,53 @@
 # about methods provided by the DSL, see https://puma.io/puma/Puma/DSL.html.
 
-# Puma can serve each request in a thread from an internal thread pool.
-# The `threads` method setting takes two numbers: a minimum and maximum.
-# Any libraries that use thread pools should be configured to match
-# the maximum value specified for Puma. Default is set to 5 threads for minimum
-# and maximum; this matches the default thread size of Active Record.
-max_threads_count = ENV.fetch("RAILS_MAX_THREADS") { 5 }
-min_threads_count = ENV.fetch("RAILS_MIN_THREADS") { max_threads_count }
-threads min_threads_count, max_threads_count
-
-# Specifies that the worker count should equal the number of processors in production.
-if ENV["RAILS_ENV"] == "production"
+rails_env = ENV.fetch("RAILS_ENV", "development")
+
+# Puma starts a configurable number of processes (workers) and each process
+# serves each request in a thread from an internal thread pool.
+#
+# As a rule of thumb, increasing the number of threads will increase how much
+# traffic a given process can handle (throughput), but due to Ruby's
+# Global VM Lock (GVL) it has diminishing returns, and will degrade the
+# response time (latency) of the application.
+#
+# The ideal number of threads per worker depends both on how much time the
+# application spends waiting for IO operations and on how much you wish to
+# to prioritize throughput over latency.
+#
+# The default is set to 3 threads as it's deemed a decent compromise between
+# throughput and latency for the average Rails application.
+#
+# Any libraries that use a connection pool or another resource pool should
+# be configured to provide at least as many connections as the number of
+# threads. This includes Active Record's `pool` parameter in `database.yml`.
+threads_count = ENV.fetch("RAILS_MAX_THREADS") { 3 }
+threads threads_count, threads_count
+
+if rails_env == "production"
+  # Specifies that the worker count should equal the number of processors in production.
+  #
+  # The number of processor is detected using various platform dependent methods, but
+  # note that this isn't very reliable, especially in containerized environments like Docker.
+  # As such, it is recommended to set it manually.
   require "concurrent-ruby"
   worker_count = Integer(ENV.fetch("WEB_CONCURRENCY") { Concurrent.physical_processor_count })
   workers worker_count if worker_count > 1
 end
 
-# Specifies the `worker_timeout` threshold that Puma will use to wait before
-# terminating a worker in development environments.
-worker_timeout 3600 if ENV.fetch("RAILS_ENV", "development") == "development"
-
 # Specifies the `port` that Puma will listen on to receive requests; default is 3000.
 port ENV.fetch("PORT") { 3000 }
 
 # Specifies the `environment` that Puma will run in.
-environment ENV.fetch("RAILS_ENV") { "development" }
-
-# Specifies the `pidfile` that Puma will use.
-if ENV["PIDFILE"]
-  pidfile ENV["PIDFILE"]
```

Comment:
Did you mean to drop support for the `PIDFILE` environment variable?

Reply:
Yes. See the PR description:
>  also removed the pidfile except for development. It's useful to prevent booting the server twice there but I don't think it makes much sense in production, especially since Puma no longer supports daemonization and instead recommend using a process monitor. And it makes even less sense in a PaaS or containerized world.

Reply:
@ezekg Yes, it is in the PR description: https://github.com/rails/rails/pull/50669#issue-2072048314

Reply:
Note: `ENV` support was explicitly kept in #50644

Reply:
Right. @rafaelfranca how much do you care about that `pidfile`? I'm really doubting it's usefulness.

Reply:
For development it is useful for tools like LSPs that can check the status of the server by checking the pidfile. I'm not sure the environment is that useful, but since it stills exists in `rails server` I think we should keep here.

Reply:
Yeah, I agree for development, that's why I left it, but I'm really doubtful about usefulness in production...

But OK, I'll leave it.

Overall it's a bit weird to have the same config in dev and prod, these environment are really considerably different, but that's another topic.

Reply:
I don't think we should add back to production, I just think we should change to be `pidfile ENV["PIDFILE"]` in development instead of `pidfile "HARDCODED PLACE"`

Reply:
Ah, I see what you mean. Yeah I can do that.